go-away/src/typescript_output/mod.rs

```diff
@@ -0,0 +1,243 @@
+use std::{fmt, fmt::Write};
+
+use indenter::indented;
+// use indoc::writedoc;
+
+mod tabify;
+
+pub use super::types::*;
+
+pub enum TypeScriptType<'a> {
+    Struct(&'a Struct),
+    NewType(&'a NewType),
+    Enum(&'a Enum),
+    Union(&'a Union),
+}
+
+impl<'a> fmt::Display for TypeScriptType<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let f = &mut tabify::tabify(f);
+        match self {
+            TypeScriptType::Struct(details) => {
+                writeln!(f, "type {} = {{", details.name)?;
+                for field in &details.fields {
+                    writeln!(indented(f), "{}", TypeScriptField(field))?;
+                }
+                writeln!(f, "}}")?;
+            }
+            TypeScriptType::NewType(details) => {
+                writeln!(
+                    f,
+                    "type {} = {}",
+                    details.name,
+                    details.inner.typescript_type()
+                )?;
+            }
+            TypeScriptType::Enum(details) => {
+                writeln!(f, "enum {} {{", details.name)?;
+                for variant in &details.variants {
+                    writeln!(
+                        indented(f),
+                        "{} = \"{}\",",
+                        variant.name,
+                        variant.serialized_name
+                    )?;
+                }
+                writeln!(f, "}}")?;
+            }
+            TypeScriptType::Union(details) => {
+                let mut union_types: Vec<String> = vec![];
+                for variant in &details.variants {
+                    union_types.push(variant.typescript_name());
+                }
+                writeln!(f, "type {} = {};", details.name, union_types.join(" | "))?;
```

Comment:
This will need fleshed out a bit more depending on `details.representation`.  Probably something like:

```rust
match details.representation {
    UnionRepresentation::Untagged => {
        // What you have now
    }
    UnionRepresentation::ExternallyTagged => {
        let mut union_types: Vec<String> = vec![];
        for variant in &details.variants {
            union_types.push(format!(r#"{{{}: {}}}"#, variant.typescript_name(), variant.typescript_name());
        }
        writeln!(f, "type {} = {};", details.name, union_types.join(" | "))?;
    },
    UnionRepresentation::InternallyTagged{tag} => {
        let mut union_types: Vec<String> = vec![];
        for variant in &details.variants {
            union_types.push(format!(r#"({{{}: {}}} & {})"#, tag, variant.typescript_name(), variant.typescript_name());
        }
        writeln!(f, "type {} = {};", details.name, union_types.join(" | "))?;
    },
    UnionRepresentation::ExternallyTagged{tag, content} => {
        let mut union_types: Vec<String> = vec![];
        for variant in &details.variants {
            union_types.push(format!(r#"{{{}: {}, {}: {}}}"#, tag, variant.typescript_name(), content, variant.typescript_name());
        }
        writeln!(f, "type {} = {};", details.name, union_types.join(" | "))?;
    }
}
```